### DISCUSSION TITLE
Detecting hardware capabilities

### DISCUSSION THREAD

**Discussion by IanRawley** [AvaloniaUI/Avalonia#13277] [Q&A] [answered]
I'm developing an app for desktop and mobile which has to be fully keyboard accessible.  This is because one of the main target user groups are people using a mobile device with physical keyboard in situations where touch interaction is impractical.  As a result, I've done a lot of focus manipulation and ensuring an appropriate control always has focus.

A side-effect of that though is that if a physical keyboard isn't present, on android the constant showing and hiding of the software keyboard has negative implications for general usability and responsiveness.  I've been searching through docs and source, but haven't had any luck finding what I'm looking for.

Are there any current APIs for detecting the presence (or absence) of specific pieces of physical hardware.  For now I'm mostly interested in keyboards, but cameras or GPS are likely to be looked at in the future.

**Comment by thevortexcloud** [comment]
I am not aware of any Avalonia specific API to do this (but it's possible I have just not seen it). But Avalonia on mobile just uses Xamarin for most of it's platform APIs. It looks like you should be able to do it assuming you can get access to the relevant APIs. Which ones you need specifically though I can't say. There are various APIs for cameras and location information.

On desktop platforms you can probably write your own somewhat easily. Although it may require dipping into native code on some platforms. On Linux it's generally as simple as checking if the device is present in the expected files.

There is however https://github.com/AvaloniaUI/Avalonia.Essentials/ which may contain what you want. I have heard people have had mixed results using it though. But I personally have never tried it.

**Comment by maxkatz6** [comment, marked as answer]
Using native APIs is your best bet in this situation:
Android has this: https://stackoverflow.com/a/57749580/5007046
iOS has this: https://developer.apple.com/documentation/gamecontroller/gckeyboard?language=objc
You can access native APIs from `net7.0-android` and `net7.0-ios` projects (shared `net7.0` doesn't have any platform APIs).

However, I don't clearly understand your problem with manual focus and soft keyboard showing on the screen. If any input control is focused, then it would be only reasonable to show soft keyboard. Or am I missing anything?

Just in case, on android you can disable soft keyboard using android manifest:
```
android:windowSoftInputMode="stateAlwaysHidden"
```

**Reply by IanRawley** [reply]
You're correct, that if an input control(where the keyboard is relevant) is focused the soft keyboard should be displayed.  My issue is that I'm currently controlling focus very strictly (an input control _always_ has focus) so as to make the app operable on a mobile device with hardware keyboard and no touchscreen or pointer device, as a significant number of my target users will be doing.  If there's no physical keyboard they need to use some sort of pointer device anyway, in which case I don't need to be such a control freak because they can just touch stuff to navigate.

EDIT: Looks like I'll have to dig into putting together a capabilities service or something and using native implementations to make the info available.
